### PR TITLE
Add Cancellation for copy log jobs #2423

### LINKED ISSUE #2423: Add Cancellation for copy log jobs
The cancellation logic should be enabled for copying logs/mnemonics. As the copy jobs are mostly generic, it it could be added for all the copy jobs.

### PR DESCRIPTION
**Pull request by robertbasti**
[//]: # (Request Template Witsml Explorer)

[//]: # (Thank you for contributing to WITSML Explorer! Before submitting this PR, please fill in the following template.)

## Fixes 

[//]: # (Write the GitHub issue number starting with #, or, for Equinor only, the Jira issue number starting with WE-)

This pull request fixes #2423

## Description

Added an cancelation capability to all copy jobs.

## Type of change

[//]: # (Mark any of the types of change that apply.)

* [ ] Bugfix
* [ ] New feature (non-breaking change which adds functionality)
* [x] Enhancement of existing functionality
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
* [ ] This change requires a documentation update

## Impacted Areas in Application

[//]: # (List general components of the application that this PR will affect)

* [x] Frontend
* [x] API
* [ ] WITSML
* [ ] Desktop
* [ ] Other (please describe)

## Checklist:

[//]: # (Please tick all the boxes or remove the ones that aren't needed and explain why)

*Communication*
* [ ] I have made corresponding changes to the documentation
* [ ] PR affects application security

*Code quality*
* [x] I have self-reviewed my code
* [x] No new warnings are generated

*Test coverage*
* [x] New code is covered by passing tests

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: Src/WitsmlExplorer.Api/Workers/Copy/CopyComponentsWorker.cs [hunk 1]
```diff
@@ -41,13 +41,14 @@ public CopyComponentsWorker(ILogger<CopyComponentsJob> logger, IWitsmlClientProv
 
         public override async Task<(WorkerResult, RefreshAction)> Execute(CopyComponentsJob job, CancellationToken? cancellationToken = null)
         {
+            // cancellationToken?.ThrowIfCancellationRequested();
```
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/Copy/CopyComponentsWorker.cs]: Remove comment

File: Src/WitsmlExplorer.Api/Workers/Copy/CopyLogDataWorker.cs [hunk 2]
```diff
@@ -195,6 +205,7 @@ private void UpdateJobProgress(CopyLogDataJob job, WitsmlLog sourceLog, WitsmlLo
             double progress = LogWorkerTools.CalculateProgressBasedOnIndex(sourceLog, copiedData);
             job.ProgressReporter?.Report(progress);
             if (job.JobInfo != null) job.JobInfo.Progress = progress;
+            if (job.JobInfo != null) job.JobInfo.Progress = progress;
```
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/Copy/CopyLogDataWorker.cs]: Remove this line (duplicate of the one above)

File: Src/WitsmlExplorer.Api/Workers/ReplaceComponentsWorker.cs
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/ReplaceComponentsWorker.cs:1]: I wouldn't expect both statuses to be 'Cancelled'. Either the first one is 'Success' and the second is 'Cancelled', or the first one is 'Cancelled', and the second line is either not added at all, or 'Not Started'.
image

**Review by eliasbruvik — CHANGES_REQUESTED** [review]

File: Src/WitsmlExplorer.Api/Workers/Copy/CopyWithParentWorker.cs
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/Copy/CopyWithParentWorker.cs:1]: Can you add some more info to the error message when you cancel CopyWithParents job? At least what kind of job was cancelled?

image

Edit: Got the same message on CopyObjectsJob aswell.

**Review by eliasbruvik — CHANGES_REQUESTED** [review]

File: Src/WitsmlExplorer.Api/Workers/Copy/CopyComponentsWorker.cs (hunk 1 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/Copy/CopyComponentsWorker.cs]: fixed

File: Src/WitsmlExplorer.Api/Workers/Copy/CopyLogDataWorker.cs (hunk 2 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/Copy/CopyLogDataWorker.cs]: fixed

File: Src/WitsmlExplorer.Api/Workers/ReplaceComponentsWorker.cs
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/ReplaceComponentsWorker.cs:1]: fixed

File: Src/WitsmlExplorer.Api/Workers/Copy/CopyWithParentWorker.cs
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/Copy/CopyWithParentWorker.cs:1]: fixed

**Review by eliasbruvik — APPROVED** [review]
*Reviewer checklist*
- [x] Read issue and PR
- [x] Pulled branch and manually tested
- [x] Verify that PR resolves issue
- [x] Reviewed the code

**Pull request merged**